TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -79,341 +119,484 @@ namespace TEN::Entities::Creatures::TR3
 		SetAnimation(item, CIVVY_ANIM_IDLE);
 	}
 
+
+	//In the future, if this function works, it could be taken to a more neutral place to be reused by other ally entities.
+	ItemInfo* CivvyFindNearTarget(short itemNumber, std::vector<GAME_OBJECT_ID>& excludedTargetList, int rangeDetection)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		ItemInfo* result = nullptr;
+
+		auto closetsDistance = FLT_MAX;
+		Vector3 distanceVector;
+		float distanceValue;
+		auto* targetCreature = ActiveCreatures[0];
+		float MaxRange = rangeDetection <= 0 ? FLT_MAX : rangeDetection;
+
+		for (int i = 0; i < ActiveCreatures.size(); i++)
+		{
+			targetCreature = ActiveCreatures[i];
+
+			//Ignore if it's itself, or a non valid Item.
+			if (targetCreature->ItemNumber == NO_ITEM || targetCreature->ItemNumber == itemNumber)
+				continue;
+
+			//Ignore if it's an entity from the Excluded Targets lists.
+			bool forbiddenTarget = false;
+			for (std::vector<GAME_OBJECT_ID>::iterator it = excludedTargetList.begin(); it != excludedTargetList.end(); ++it)
+			{
+				if (g_Level.Items[targetCreature->ItemNumber].ObjectNumber == *it)
+					forbiddenTarget = true;
+			}
+			if (forbiddenTarget)
+				continue;
+
+			//If it's closer than other entity, choose this one.
+			auto& currentItem = g_Level.Items[targetCreature->ItemNumber];
+
+			distanceValue = Vector3i::Distance(item.Pose.Position, currentItem.Pose.Position);
+
+			if (distanceValue < closetsDistance && distanceValue < MaxRange)
+			{
+				closetsDistance = distanceValue;
+				result = &currentItem;
+			}
+		}
+		return result;
+	}
+
 	void CivvyControl(short itemNumber)
 	{
 		if (!CreatureActive(itemNumber))
 			return;
 
-		auto* item = &g_Level.Items[itemNumber];
-		auto* creature = GetCreatureInfo(item);
+		auto& item = g_Level.Items[itemNumber];
+		auto& object = Objects[item.ObjectNumber];
+		auto& creature = *GetCreatureInfo(&item);
 
 		short angle = 0;
 		short tilt = 0;
-		auto extraHeadRot = EulerAngles::Zero;
-		auto extraTorsoRot = EulerAngles::Zero;
 
-		if (item->BoxNumber != NO_BOX && (g_Level.Boxes[item->BoxNumber].flags & BLOCKED))
+		int targetAngle = 0;
+		int targetDistance = 0;
+		auto jointHeadRot = EulerAngles::Zero;
+		auto jointTorsoRot = EulerAngles::Zero;
+
+		if (item.BoxNumber != NO_BOX && (g_Level.Boxes[item.BoxNumber].flags & BLOCKED))
 		{
-			DoDamage(item, 20);
-			DoLotsOfBlood(item->Pose.Position.x, item->Pose.Position.y - (GetRandomControl() & 255) - 32, item->Pose.Position.z, (GetRandomControl() & 127) + 128, GetRandomControl() << 1, item->RoomNumber, 3);
+			DoDamage(&item, 20);
```

Comment:
you can kill him directly with `INT_MAX` since it won't change anything. also check if he is not already dead with `item.HitPoints > 0`

Reply:
Done